tools/omnissm/host/register/initialize.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2018 Capital One Services, LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+
+You may obtain a copy of the License at
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// SSMAgentRegistration Sourced from /var/lib/amazon/ssm/registration
+type SSMAgentRegistration struct {
+	ManagedInstanceID string
+	Region            string
+}
+
+// LinuxSSMRegistrationPath Linux Path to Agent Registration State
+const LinuxSSMRegistrationPath = "/var/lib/amazon/ssm/registration"
+
+// IdentityURL ec2 metadata server instance identity document
+const IdentityURL = "http://169.254.169.254/latest/dynamic/instance-identity/document"
+
+// SignatureURL RSA SHA256 Signature of identity document
+const SignatureURL = "http://169.254.169.254/latest/dynamic/instance-identity/signature"
+
+// FetchContents Retrieve contents of URL
+func FetchContents(uri string) ([]byte, error) {
+	response, err := http.Get(uri)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+	return body, nil
+}
+
+func main() {
+	// If no ssm-agent installation, exit
+	ssmCmdPath, err := exec.LookPath("amazon-ssm-agent")
+	if err != nil {
+		fmt.Println("Could not find amazon-ssm-agent binary")
+		return
+	}
+	// If we have a current registration for hybrid mode, exit
+	agentRaw, err := ioutil.ReadFile(LinuxSSMRegistrationPath)
+	if err == nil {
+		agentReg := SSMAgentRegistration{}
+		err = json.Unmarshal(agentRaw, &agentReg)
+		if err == nil {
+			if strings.HasPrefix(agentReg.ManagedInstanceID, "mi-") {
+				fmt.Println("SSM Agent registered already", agentReg.ManagedInstanceID)
+				return
+			}
+			fmt.Println("SSM Agent overriding registration", agentReg.ManagedInstanceID)
+		}
+	}
+
+	identity, err := FetchContents(IdentityURL)
+	if err != nil {
+		panic(err)
```

Comment:
Earlier versions of Go and/or users with particular environment variables set will get the full stack trace for every Goroutine when panicking.  Because of this it would be better to `log.Fatal` or similar instead of panic since this will produce the undesired effect for command-line users by giving them a wall of text rather than a single line error message.  This practice should be observed throughout all main function and panics should be avoided unless truly called for.

Reply:
we're pretty early in dev (all go 1.10+), so i kinda of want the panics here to get as much debug info as possible, we'll be setting up the systemd/upstart job files for these. That said trying to keep any panics at the top level entry points (main / lambda handler), log.Fatal looks like we just get a single error line and exit 1 . The intent on this script isn't actually end user execution, its being driven by ssm agent across a fleet with error collection.